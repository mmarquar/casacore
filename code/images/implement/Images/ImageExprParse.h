//# LatticeParse.h: Classes to hold results from lattice expression parser
//# Copyright (C) 1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#if !defined(AIPS_LATTICEPARSE_H)
#define AIPS_LATTICEPARSE_H


//# Includes
#include <trial/Lattices/LatticeExpr.h>
#include <aips/Mathematics/Complex.h>
#include <aips/Utilities/String.h>
#include <aips/Utilities/DataType.h>

//# Forward Declarations


// <summary>
// Class to hold values from lattice expression parser
// </summary>

// <use visibility=export>

// <reviewed reviewer="" date="" tests="">
// </reviewed>

// <prerequisite>
//# Classes you should understand before using this one.
//  <li> <linkto class=LatticeExpr>LatticeExpr</linkto>
// </prerequisite>

// <etymology>
// LatticeParse is the class used to parse a lattice command.
// </etymology>

// <synopsis> 
// LatticeParse is used by the parser of lattice expression statements.
// The parser is written in Bison and Flex in files LatticeGram.y and .l.
// The statements in there use the routines in this file to act
// upon a reduced rule.
// <p>
// The class contains some values used by the parsing process.
// The only function to be used by a user is the static
// function LatticeParse::command which parses an expression.
// It returns a <linkto class=LatticeExprNode>LatticeExprNode</linkto>
// object containing the expression represented as a tree.
// The object can be used as a <src>Lattice(Expr)</src> in other operations.
// </synopsis> 

// <example>
// <srcblock>
//    LatticeExpr<Double> expr ("a + sin(b)");
//    ArrayLattice<Double> arr(expr.shape());
//    arr.copyData (expr);
// </srcblock>
// Line 1 creates a LatticeExpr object for the given expression. Note that
// <src>a</src> and <src>b</src> are names of lattice files (e.g. PagedImage).
// <br> Line 2 creates an ArrayLattice with the same shape as the expression
// (which is the shape of lattice a (and b)).
// <br> Line 3 copies the result of the expression to the ArrayLattice.
// </example>

// <motivation>
// It is necessary to be able to give a lattice expression command in ASCII.
// This can be used in glish to operate on lattices/images.
// </motivation>

//# <todo asof="$DATE:$">
//# A List of bugs, limitations, extensions or planned refinements.
//# </todo>


class LatticeParse
{
public:

    // Parse the given command.
    // It will open all lattices needed.
    // It returns the resulting lattice expression.
    static LatticeExprNode command (const String& str);

    // Construct a literal object for the given type.
    // <group>
    LatticeParse (Int value);
    LatticeParse (Float value);
    LatticeParse (Double value);
    LatticeParse (const Complex& value);
    LatticeParse (const DComplex& value);
    LatticeParse (const String& value);
    // </group>

    // Make a LatticeExprNode for a function.
    // <group>
    LatticeExprNode makeFuncNode () const;
    LatticeExprNode makeFuncNode (const LatticeExprNode& arg1) const;
    LatticeExprNode makeFuncNode (const LatticeExprNode& arg1,
				  const LatticeExprNode& arg2) const;
    // </group>

    // Make a LatticeExprNode object for the lattice.
    LatticeExprNode makeLatticeNode() const;

    // Make a LatticeExprNode object for the literal.
    LatticeExprNode makeLiteralNode() const;

    // Set the static node object (used by the .y file).
    static void setNode (const LatticeExprNode& node)
        { theirNode = node; }

private:
    //# A 'global' node object to hold the resulting expression.
    static LatticeExprNode theirNode;

    DataType itsType;
    Int      itsIval;              //# integer literal
    Float    itsFval;              //# Float literal
    Double   itsDval;              //# Double literal
    Complex  itsCval;              //# Complex literal
    DComplex itsDCval;             //# DComplex literal
    String   itsSval;              //# lattice name; function name
};


#endif
