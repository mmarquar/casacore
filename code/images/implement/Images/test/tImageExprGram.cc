//# tImageExprGram.cc: Test program for image expression parser
//# Copyright (C) 1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <trial/Images/ImageExprParse.h>
#include <trial/Lattices/LatticeExprNode.h>
#include <trial/Lattices/PagedArray.h>
#include <trial/Lattices/ArrayLattice.h>
#include <aips/Arrays/Array.h>
#include <aips/Arrays/ArrayLogical.h>
#include <aips/Lattices/IPosition.h>
#include <aips/Containers/Block.h>
#include <aips/Mathematics/Constants.h>
#include <aips/Inputs/Input.h>
#include <aips/Exceptions/Error.h>
#include <iostream.h>


main (int argc, char *argv[])
{
 try {
    Input inp(1);
    inp.Version(" ");
    inp.Create("nx", "2", "Number of pixels along the x-axis", "int");
    inp.Create("ny", "2", "Number of pixels along the y-axis", "int");
    inp.ReadArguments(argc, argv);

    const uInt nx=inp.GetInt("nx");
    const uInt ny=inp.GetInt("ny");

    Double aVal = 0.0;
    Double bVal = 1.0;
    Double cVal = 2.0;
    Double dVal = 3.0;
    Double eVal = 4.0;
    Double fVal = 5.0;
    Double gVal = 6.0;
    Double hVal = 7.0;
    Bool aBoolVal = False;
    Bool bBoolVal = False;
    ArrayLattice<Double> a(IPosition (2,nx,ny));
    ArrayLattice<Bool> aBool(IPosition (2,nx,ny));
    a.set(aVal);
    aBool.set(aBoolVal);
    {
	PagedArray<Double> b(IPosition (2,nx,ny), "b");
	PagedArray<Double> c(IPosition (2,nx,ny), "c");
	PagedArray<Double> d(IPosition (2,nx,ny), "d");
	PagedArray<Double> e(IPosition (2,nx,ny), "e");
	PagedArray<Double> f(IPosition (2,nx,ny), "f");
	PagedArray<Double> g(IPosition (2,nx,ny), "g");
	PagedArray<Double> h(IPosition (2,nx,ny), "h");
	PagedArray<Bool> bBool(IPosition (2,nx,ny), "bBool");
	
	b.set(bVal);
	c.set(cVal);
	d.set(dVal);
	e.set(eVal);
	f.set(fVal);
	g.set(gVal);
	h.set(hVal);
	bBool.set(bBoolVal);
    }

    Array<Double> aArr(a.shape());
    Array<Bool> aBoolArr(aBool.shape());

    Bool foundError = False;

  {
    cout << endl;
    cout << "Expr:  a = 1" << endl;
    a.copyData(ImageExprParse::command ("1.0"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = 1.0;
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = b" << endl;
    a.copyData(ImageExprParse::command ("b"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = bVal;
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result  << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = sin(\\c)" << endl;
    a.copyData(ImageExprParse::command ("sin(\\c)"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = sin(cVal);
    if (! allEQ (aArr, sin(cVal))) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }

  {
    cout << "Expr:  a = 'c'+2" << endl;
    a.copyData(ImageExprParse::command ("'c'+2"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = cVal+2;
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = b+('c'+d)     (using $n notation)" << endl;
    PagedArray<Double> b("b");
    PagedArray<Double> c("c");
    PagedArray<Double> d("d");
    Block<LatticeExprNode> temps(3);
    temps[0] = LatticeExprNode(b);
    temps[1] = LatticeExprNode(c);
    temps[2] = LatticeExprNode(d);
    a.copyData(ImageExprParse::command ("$1+($2+$3)", temps));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = bVal + cVal + dVal;
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = 3.5*b + cos('c')-10/min('c',d)*-e*log(b)-pi()" << endl;

    a.copyData( ImageExprParse::command
               ("(3.5*b) + (cos('c')) - (10/min('c',d)*(-e)*log(b)) - (pi()) "));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = 3.5*bVal + cos(cVal) -
	            10/min(cVal,dVal)*-eVal*log(bVal) - C::pi;
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = (b+'c'-d/2.0*-b) + pi()" << endl;

    a.copyData(ImageExprParse::command ("(b+'c'-d/2.0*-b)+pi()"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = (bVal + cVal - dVal / 2*-bVal) + C::pi;
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = pow('c',d)" << endl;

    a.copyData(ImageExprParse::command ("pow('c',d)"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = pow(cVal,dVal);
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = fmod('c'*2,d)" << endl;

    a.copyData(ImageExprParse::command ("fmod('c'*2,d)"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = fmod(cVal*2,dVal);
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = pow('c',2.3)" << endl;

    a.copyData(ImageExprParse::command ("pow('c',2.3)"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = pow(cVal,2.3);
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = cos(b)" << endl;

    a.copyData(ImageExprParse::command ("cos(b)"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = cos(bVal);
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = cos(sin(b))" << endl;

    a.copyData (ImageExprParse::command ("cos(sin(b))"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = cos(sin(bVal));
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = ntrue(b>=1)" << endl;

    a.copyData(ImageExprParse::command ("ntrue(b>=1)"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = a.shape().product();
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  aBool = T||F" << endl;

    aBool.copyData(ImageExprParse::command ("T||F"));
    aBool.getSlice(aBoolArr, IPosition(aBoolArr.ndim(),0), 
		   aBoolArr.shape(), IPosition(aBoolArr.ndim(),1));
    Bool result = ToBool(True||False);
    if (! allEQ (aBoolArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aBoolArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  aBool = !bBool" << endl;

    aBool.copyData(ImageExprParse::command ("!bBool"));
    aBool.getSlice(aBoolArr, IPosition(aBoolArr.ndim(),0), 
		   aBoolArr.shape(), IPosition(aBoolArr.ndim(),1));
    Bool result = ToBool(!bBoolVal);
    if (! allEQ (aBoolArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aBoolArr.ac() << endl;
	foundError = True;
    }
  }
  {
    cout << "Expr:  a = sum(e)/nelements(b) + min('c') + max('c', mean('c'+d))" << endl;

    a.copyData(ImageExprParse::command
	          ("sum(e)/nelements(b) + min('c') + max('c', mean('c'+d))"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = eVal+cVal+(cVal+dVal);
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }
  {
    {
	PagedArray<Double> ap (a.shape(), "a");
	ap.copyData (a);
    }
    cout << "Expr:  a = min(a+10,5)" << endl;

    a.copyData(ImageExprParse::command ("min(a+10,5)"));
    a.getSlice(aArr, IPosition(aArr.ndim(),0), 
	       aArr.shape(), IPosition(aArr.ndim(),1));
    Double result = 5;
    if (! allEQ (aArr, result)) {
	cout << "Result should be " << result << endl;
	cout << "Result is " << aArr.ac() << endl;
	foundError = True;
    }
  }

  cout << endl;
  if (foundError) {
     return 1;
  }

 } catch (AipsError x) {
    cerr << "aipserror: error " << x.getMesg() << endl;
    return 1;
 } end_try;
 
 return 0;
}
