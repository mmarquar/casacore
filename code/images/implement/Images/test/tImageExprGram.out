>>>
tImageExprGram: Version 1.0
<<<

ImageExprParse: 'xxx' is an unknown lattice or image or it is an unqualified region
Scanned so far: xxx

ImageExprParse: 'a1' is an unknown lattice, image, or region
Scanned so far: b/a1

ImageExprParse: 'a1' is an unknown lattice or image or it is an unqualified region
Scanned so far: a1/
Image Expression: Parse error at or near ''
Scanned so far: b/b*

3-argument function min is unknown
Scanned so far: min(b,b,b)

Expr:  a = 1
Expr:  a = b
Expr:  a = sin(\c)
Expr:  a = 'c'+2
Expr:  a = b+('c'+d)[region]     (using $n notation)
Expr:  a = nelements(b[$region]
Expr:  a = nelements(b[$region1 || $region2) - length(b[$region1],1)
Expr:  a = 3.5*b + cos('c')-10/min('c',d)*-e*log(b)-pi()
Expr:  a = (b+'c'-d/2.0*-b) + pi()
Expr:  a = pow('c',d)
Expr:  a = fmod('c'*2,d)
Expr:  a = pow('c',2.3)
Expr:  a = cos(b)
Expr:  a = cos(sin(b))
Expr:  a = ntrue(b>=1)
Expr:  aBool = T||F
Expr:  aBool = !bBool
Expr:  a = sum(e)/nelements(b) + min('c') + max('c', mean('c'+d))
Expr:  a = sum(e[bBool])
Expr:  a = sum((e + sum(f[b>0]*c)/d)[!bBool])
Expr:  a = min(a+10,5)

